### ISSUE TITLE
Sortupdate event index wrong

### ISSUE THREAD

**Issue by lukasoppermann** [lukasoppermann/html5sortable#381]
The values returned in the sortupdate are not always correct.

**Comment by mfeherpataky** [comment]
The problem is with startSortableIndex, right?
I will look at it in the next few days.

**Comment by lukasoppermann** [comment]
Exactly. Thanks!

By the way, I would like to move all the event functions outside of the sortable so that they are only accessing things that are passed into them. If you while working on this issue want to work towards that, that would be neat.

**Comment by mfeherpataky** [comment]
warning - long :grey_exclamation:

While there is a simple fix - just a simple refresh of sortable items cache (var items) on dragstart.
It only works for items moved within a single sortable list. The `sortupdate` is fired off on `dragend` or `drop` in a scope of a sortable item and depends on some local cache-like variables which are set on `dragstart`.

As you you are suggesting - a proper fix probably requires to gut the event function completely from any one sortable. In some cases the event function would need to get values from multiple sortables. 

The cases I see:

a) Item moved within the same list (data from a single sortable list)
b) Item moved from another sortable list where we would want to track both lists (need to pass data from both sortable lists)
c) Item moved form another sortable list where items are copied from, so no need to track both lists. (single list?)

---

The detail object has the following:
- `item` - the element we moved around
- `startParent` - starting List (works as it is held outside of a sortable scope)
- `endParent` - target List
-  :warning: `startIndex` - starting Index within the endParent list (that is set on dragstart in the endParent sortable, which is not fired when dropping in the endParent)
- `endIndex` - target Index within the endParent list
- :warning: `startSortableIndex` - starting Index in the endParent list (that is set on dragstart in the endParent sortable, which is not fired when dropping in the endParent) 
- `endSortableIndex` - target Index in the endParent list

What do you expect to get with startIndex and startSortableIndex in case (b) & (c) - when item is moved across lists? I would expect indexes in the startParent lists, which is not the case.

Going one step forward - there are three serialised lists:
- :warning: `oldStartList` - currently points to the pre-update state of endParent list... Not what I think is expected. May also be undefined as `dragstart` was not fired in the endParent
- `newStartList` - points to post-update state of the startParent list (that's OK)
- `newEndList` - points to post-update state of endParent list (that's OK)

----

**TL;DR I would suggest:**
- a small refactoring of local variables maybe with some further renaming to clarify scope - setting them at the top of the sortable with ES6 let styles and some comments,
- move CustomEvent triggering code to a dedicated file "customEventDispatcher" or some other name you like with selected values passed from each native drag-n-drop events. I would move most of the custom event dispatching logic there as well to cleanup the main function of html5sortable.ts

Do you see other options? Does it make sense what I say?

**Comment by lukasoppermann** [comment]
Hey @mfeherpataky, I agree with the issues you point out & ideas. Very nice. Thanks for planing and suggesting changes beforehand. 😉 👍 

So changes would be:

**First PR:**
- small refactoring of local variables with some renaming
-> which fixes `startIndex`, `startSortableIndex` & `oldStartList` (Also when moving from one list to another)

**Second PR (or better following PRs):**
- move CustomEvent triggering code to a dedicated file
->  moving most of the custom event dispatching logic there as well sounds good 👍 
-> however, I would suggest one file per Event, e.g. `sortStartEvent.ts` (does that make sense to you, small files are much easier to scan and work with)

**Keep in mind:**
I am currently working on refactoring the way con configuration, data & more is stored in https://github.com/lukasoppermann/html5sortable/pull/369

In the first merged version it will only change `config`/options, however it will be expanded to include data & placeholders afterwards. This should make working with data much easier, as it can be easily accessed by using the `store(sortableElement)` functions e.g. `store(sortableElement).getConfig('acceptFrom')`. Maybe this helps/changes something & maybe if you extract/rewrite stuff you can already prepare it to be more easily substituted by this function.

**Comment by lukasoppermann** [comment]
Will this fix #281 as well?

**Comment by mfeherpataky** [comment]
Yeap it will fix it.
Though it is no longer an 'oldIndex', but rather 'startSortableIndex'.

Do you have any better names for these values?
I am not convinced that the 'start/end' pattern I proposed makes them any clearer.

Perhaps simplifying them?

- `startingIndex` - index relocated item started with
- `index` - the latest item index (just the index)

--

- `startingParent` - parent list where relocated item resided at the start
- `parent` - the latest parent list of the item (just the parent)

--

- `startingSortableIndex` - index relocated item started with (scoped to sortable items in 'startingParent')
- `sortableIndex` - latest index (scoped to sortable items in 'parent')

--

- `startingOriginalList` - serialised list of 'startingParent' AS-WAS before any relocation took place
- `startingList` - serialised list of 'startingParent' AS-IS after the relocation
- `list` - serialised list of the current 'parent'

---

If I get your green-light I would use the same names for the local variables we will use to fix #281 & #381 so that we move the first PR out of the way.

**Comment by lukasoppermann** [comment]
Hey @mfeherpataky, I am definitely in favor of simplifying and U like your ideas, but I am wondering if we can make it even more explicit.

We could use `from` and `to` or `source`/`origin` and `destination`? An maybe that could be objects, e.g.

```javascript
{
origin: {
  elementIndex: 1,
  index: 0,
  container: HTMLElement (sortable = parent but like in serialize function),
  itemsBeforeUpdate: Array<HTMLElement>,
  items: Array<HTMLElement>
},
destination: {
  elementIndex: 3,
  index: 3,
  container: HTMLElement (sortable = parent but like in serialize function),
  itemsBeforeUpdate: Array<HTMLElement>,
  items: Array<HTMLElement>
},
item: HTMLElement
}
```

This way the user can easily do `destination.index` to get  a value.

We should also add the item that was moved or copied, maybe just as `item`.

What do you think (this is just an idea, so please let me know any concerns). I also changed the sortableIndex to index, and index to elementIndex. My reasoning is that people will probably want the index of the sortable items (e.g. if a title element is in the list, like an `<lh>`) and thus the index of only the sortables should be easier and more shown as the default value.

**Comment by mfeherpataky** [comment]
`origin/destination` differentiation definitely clears things up :+1:

I would propose one modification and place 'elementIndex' only if there are non-sortable items in either of the origin/destination lists. Would it make sense?

**Comment by lukasoppermann** [comment]
Hey @mfeherpataky I see where you are coming from, but I think this makes it more complicated for the user. When building APIs I have learned that it is always best to stick with consistent return values.

Thus I would like to keep `elementIndex` in the response no matter what. This way the user knows that `index` and `elementIndex` will reliably be present. They might be both the same, but they are definitely there. Does this make sense to you?

**Comment by mfeherpataky** [comment]
Sure! It is your baby 😄

**Comment by lukasoppermann** [comment]
However, this does not mean that I always have the better idea. 😉 But I take it that you think this will be fine? Otherwise, if you do have concerns, please let me know.

**Comment by xGouley** [comment]
This bug should be related to another :
If you have a single sortable list, sortupdate is triggered if first element is moved after the last,
BUT is not triggered when the last item is moved a fairst place (because start/end indexes are the same ?)

In a wider case, when you have a single sortable list, moving an item to one position up (to decrease its index by one position) won't trigger the sortupdate.

**Comment by mfeherpataky** [comment]
@xGouley did you pull the attached PR and tested your cases?

From what I remember - that's all covered. Give it a go, I think you should be OK. 
Do test and let us know.

I will come back to this code in a day or two, clean it up some more and maybe it would be OK for a merge.

**Comment by lukasoppermann** [comment]
Hey @xGouley can you please verify & confirm that this issue you described is fixed?

**Comment by lukasoppermann** [comment]
Closed
